cmd/app/http.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package app
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/pkg/errors"
+	"github.com/sigstore/fulcio/pkg/api"
+	gw "github.com/sigstore/fulcio/pkg/generated/protobuf"
+	legacy_gw "github.com/sigstore/fulcio/pkg/generated/protobuf/legacy"
+	"github.com/sigstore/fulcio/pkg/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
+)
+
+type httpServer struct {
+	*http.Server
+	httpServerEndpoint string
+}
+
+func extractOIDCTokenFromAuthHeader(ctx context.Context, req *http.Request) metadata.MD {
+	token := strings.Replace(req.Header.Get("Authorization"), "Bearer ", "", 1)
+	return metadata.Pairs(api.MetadataOIDCTokenKey, token)
+}
+
+func createHTTPServer(ctx context.Context, serverEndpoint string, grpcServer, legacyGRPCServer *grpcServer) httpServer {
+	mux := runtime.NewServeMux(runtime.WithMetadata(extractOIDCTokenFromAuthHeader),
+		runtime.WithForwardResponseOption(setResponseCodeModifier))
+
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+	if err := gw.RegisterCAHandlerFromEndpoint(ctx, mux, grpcServer.grpcServerEndpoint, opts); err != nil {
+		log.Logger.Fatal(err)
+	}
+
+	if legacyGRPCServer != nil {
+		endpoint := fmt.Sprintf("unix:%v", legacyGRPCServer.grpcServerEndpoint)
+		if err := legacy_gw.RegisterCAHandlerFromEndpoint(ctx, mux, endpoint, opts); err != nil {
+			log.Logger.Fatal(err)
+		}
+	}
```

Comment:
What determines if `legacyGRPCServer` is set , does this bubble up from config?

Reply:
it doesn't bubble up from config (yet), but if we wanted to deprecate and turn that off we could just set it to nil

Reply:
gotcha, I was thinking there might be a flag/config leveraged that would decide on what runs (restAPI or gPRC (or even both)).

Reply:
hmm, interesting thought. let me noodle on that for a bit.